_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,353 @@
+    case _ => None
+  }
+}
+```
+
+Notice that trying to extend Reservation on another file, possibly bypassing our <strong>apply</strong> method, will not work.
+
+As an alternative, one can hide the case class altogether, i.e. not only the constructor but also the name (so that we cannot use it as a type). To that end one defines a trait which exposes the implementation that we want.
+
+
+```scala
+sealed trait Reservation[SeatInfo <: Seat] {
+  val userTicket: Ticket[SeatInfo]
+  val guestTicket: Ticket[SeatInfo]
+}
+```
+
+```scala
+object Reservation {
+
+  private case class ReservationImpl[SeatInfo <: Seat](
+    userTicket: Ticket[SeatInfo],
+    guestTicket: Ticket[SeatInfo]
+  ) extends Reservation[SeatInfo]
+
+  def apply[SeatedInfo <: Seat](userTicket: Ticket[SeatedInfo], guestTicket: Ticket[SeatedInfo]): Option[Reservation[SeatedInfo]] = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => Some(ReservationImpl(userTicket, guestTicket))
+    case (VipTribuneSeat(_), VipTribuneSeat(_)) => Some(ReservationImpl(userTicket, guestTicket))
+    case (_: Standing.type , _: Standing.type ) => Some(ReservationImpl(userTicket, guestTicket))
+    case _ => None
+  }
+}
+```
+
+Notice the relevance of declaring the trait sealed. It prevents the user from extending the trait by another class on some other file, which would allow him to bypass the factory method <strong>apply</strong> above. Equally important is to have ReservationImpl as private so that it cannot be extended on client code. In fact, I guess we could even ignore ReservationImpl altogether and instantiate an anonymous class instead.
+
+
+This approach has a disadvantage concerning type erasure which also plagues the next approach using generalized type constraints. We will address the issue there.  
+
+# 3. Using generalized type constrains
+
+This is the most novel of the approaches and, as far as I know, could not be achieved by Java, contrary to the remaining solutions.  
+
+Let us recall the original problem. On:  
+
+```scala
+case class Reservation[SeatInfo <: Seat](
+  userTicket: Ticket[SeatInfo],
+  guestTicket: Ticket[SeatInfo]
+)
+```
+the compiler can impose <code>SeatInfo</code> to have as upper bound the type <code>Seat</code>. It cannot however impose <code>SeatInfo</code> to be one of the sub-types exclusively. Meaning that upon userTicket: <code>Ticket[VipTribune]</code> and guestTicket: <code>Ticket[NormalSeat]</code>, both <code>VipTribune</code> and <code>NormalSeat</code> can pretend to be type <code>Seat</code> which would result in userTicket: <code>Ticket[Seat]</code> and guestTicket: <code>Ticket[Seat]</code> which in turn conforms with the upper bound constraint (Reservation[SeatInfo <: Seat]). The issue really is that a given type A is a subtype of itself.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />      
```

Comment:
can you put this image in public folder of the repo?

Reply:
See first comment